containers/src/Kokkos_DualView.hpp

```diff
@@ -480,12 +498,9 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
   ///   the data in either View.  You must manually mark modified data
   ///   as modified, by calling the modify() method with the
   ///   appropriate template parameter.
-  template <class Device>
-  void sync(const typename std::enable_if<
-                (std::is_same<typename traits::data_type,
-                              typename traits::non_const_data_type>::value) ||
-                    (std::is_same<Device, int>::value),
-                int>::type& = 0) {
+  // deliberately passing args by value as they're used multiple times
+  template <class Device, class... Args>
```

Comment:
After discussing this with @jrmadsen, the safest (and likely better performance) way to receive a parameter pack being used multiple times is by const l-value reference.

The general problem is that the callee can take parameters by non-const reference and modify them (this is obvious in the case of passing unnamed temporaries by non-const r-value reference, but the problem is more general than that), which we really don't want if we are passing them multiple times.  Better to receive by const reference (and the distinction between const l-value reference and const r-value reference isn't worth doing) so that the callees can't modify their values (`const_cast` and `mutable` notwithstanding, of course).  Plus, copies are only made if the callee does it (either implicitly with pass by value or explicitly in their code).

In other words, it should be written as:
```
template <class Device, class... Args>
void sync_impl(std::true_type, Args const&... args) {
```

Reply:
Thanks for the explanation, done via e198274